lib/tbot/testhelpers/srv.go

```diff
@@ -49,35 +59,53 @@ func DefaultConfig(t *testing.T) *config.FileConfig {
 		Proxy: config.Proxy{
 			Service: config.Service{
 				EnabledFlag:   "true",
-				ListenAddress: mustGetFreeLocalListenerAddr(t),
+				ListenAddress: newListener(t, listenerProxySSH, &fds),
 			},
-			WebAddr:    mustGetFreeLocalListenerAddr(t),
-			TunAddr:    mustGetFreeLocalListenerAddr(t),
+			WebAddr:    newListener(t, listenerProxyWeb, &fds),
+			TunAddr:    newListener(t, listenerProxyTunnel, &fds),
 			PublicAddr: []string{"proxy.example.com"},
 		},
 		Auth: config.Auth{
 			Service: config.Service{
 				EnabledFlag:   "true",
-				ListenAddress: mustGetFreeLocalListenerAddr(t),
+				ListenAddress: newListener(t, listenerAuthSSH, &fds),
 			},
 		},
 	}
+
+	return fc, fds
 }
 
-func mustGetFreeLocalListenerAddr(t *testing.T) string {
+func newListener(t *testing.T, ty string, fds *[]service.FileDescriptor) string {
+	t.Helper()
+
 	l, err := net.Listen("tcp", "127.0.0.1:0")
 	require.NoError(t, err)
 	defer l.Close()
-	return l.Addr().String()
+	addr := l.Addr().String()
+
+	lf, err := l.(*net.TCPListener).File()
```

Comment:
I'm assuming that the magic in this PR is keeping the file descriptors for each port around for the duration of the test, but I don't understand why this would be so.

If that's the case, can you add a few lines of commentary explaining how?

Reply:
Yep, the idea is that we create a listener on `127.0.0.1:0`, which gives us a listening tcp socket on localhost with a port that's randomly chosen by the OS from the ones that are not in use, and then we keep that listening socket open (duplicating it into a `*os.File` and closing the original still counts) and we pass it to teleport as part of the fd passing mechanism that's used during SIGHUP restarts (where file descriptors are "exported" and passed to the child process through inheritance) and in-process restarts (where they're just passed in through the `FileDescriptors` field in `Config` like we do here). The end result is that the listening socket is never closed, and nothing else in the system (including ourselves) can mess with it.

A similar approach (because it would probably end up reusing the same mechanism anyway - internally we really like grabbing other services' `listen_addr` to connect to, so we'd need to decide what the actual address *is* before the other service might've started) would be to support `:0` in configuration files, but I can't really think of a reason why you'd ever use that in regular teleport, we'd have to decide what that means in the context of restarts (do you open a new listening socket on a different port, or do you reuse the inherited listener?), and in general it seems like a pitfall we don't want to leave open for our actual users.

Added a comment explaining this as well.